crypto/x509/by_store.c

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/store.h>
+#include "internal/cryptlib.h"
+#include "internal/x509_int.h"
+#include "x509_lcl.h"
+
+/* Generic object loader, given expected type and criterion */
+static int cache_objects(X509_LOOKUP *lctx, const char *uri,
+                         X509_LOOKUP_TYPE type,
+                         const OSSL_STORE_SEARCH *criterion,
+                         int depth)
+{
+    int ok = 0;
+    OSSL_STORE_CTX *ctx = NULL;
+    int expect = 0;
+    X509_STORE *xstore = X509_LOOKUP_get_store(lctx);
+
+    if (type == X509_LU_X509)
+        expect = OSSL_STORE_INFO_CERT;
+    else if (type == X509_LU_X509)
+        expect = OSSL_STORE_INFO_CRL;
+    if (expect == 0) {
+        X509err(X509_F_CACHE_OBJECTS, X509_R_WRONG_LOOKUP_TYPE);
+        return 0;
+    }
+
+    if ((ctx = OSSL_STORE_open(uri, NULL, NULL, NULL, NULL)) == NULL)
+        return 0;
+    OSSL_STORE_expect(ctx, expect);
+    if (criterion != NULL)
+        OSSL_STORE_find(ctx, criterion);
+
+    while (!OSSL_STORE_eof(ctx) && !OSSL_STORE_error(ctx)) {
+        OSSL_STORE_INFO *info = OSSL_STORE_load(ctx);
+        int infotype = info == 0 ? 0 : OSSL_STORE_INFO_get_type(info);
+
+        ok = 0;
+
+        if (info == NULL)
+            continue;
+
+        if (infotype == OSSL_STORE_INFO_NAME) {
+            /*
+             * This is an entry in the "directory" represented by the current
+             * uri.  if |depth| allows, dive into it.
+             */
+            if (depth == 0)
+                ok = cache_objects(lctx, OSSL_STORE_INFO_get0_NAME(info),
+                                     type, criterion, depth - 1);
+        } else {
+            /*
+             * We know that X509_STORE_add_{cert|crl} increments the object's
+             * refcount, so we can safely use OSSL_STORE_INFO_get0_{cert,crl}
+             * to get them.
+             */
+            switch (infotype) {
+            case OSSL_STORE_INFO_CERT:
+                ok = X509_STORE_add_cert(xstore,
+                                         OSSL_STORE_INFO_get0_CERT(info));
+                break;
+            case OSSL_STORE_INFO_CRL:
+                ok = X509_STORE_add_crl(xstore,
+                                        OSSL_STORE_INFO_get0_CRL(info));
+                break;
+            }
+        }
+
+        OSSL_STORE_INFO_free(info);
+        if (!ok)
```

Comment:
Logic with "ok" flag is not correct. Method X509_STORE_add* return false(zero) in (1) case of error or if (2) entity is already is store. The second one is not error.

Reply:
I have no idea where you get that from.  If you look at them, it's very clear that a returned 0 is intended to mean error, 1 is meant to mean success.

https://github.com/openssl/openssl/blob/ffa9bff8a2024bc8e6322be2cbff7d17fb8d828d/crypto/x509/x509_lu.c#L364-L380

The internal function `x509_store_add` returns *1* if the object could be found in the store, not 0.